### PR TITLE
conditions: allow using math in compare-conditions

### PR DESCRIPTION
**Pull request by bloerwald**
This changeset allows writing scripts with more complex math, like #10, but with more operations

```
test(self.speed > clamp(1, 2, 3)) [ self.speed > clamp(1, 2, 3) ]
test(self.hp > cond(false, 1200, 10)) [ self.hp > cond(false, 1200, 10) ]
test(self.hp > ceil(1.5*floor (3/2))) [self.hp > ceil(1.5*floor(3/2)) ]
test(self.hp > 3%2) [ self.hp > 3%2 ]
test(self.hp > 1*2/3-4+5) [self.hp > 1*2/3-4+5]

use(Call Lightning:204)
use(Ion Cannon:209) [ enemy.hp <= (732 * 1.25) / 2 ]
use(Metal Fist:384)
```

### CODE REVIEW

**Comment by erigul** [comment]
Would we want to return the floored (or rounded) value of the expression instead of the actual value?
`(732 * 1.25) / 2 = 457.5` but I think in this case pets actually hit for `457`.
I guess this only matters in rare cases where the enemy pet has exactly 457 health. I'm not sure if the actual attack value is floored or rounded to closest int, but I think at least the logic to the 'hp' function should consider this.

**Review by axc450 — APPROVED** [review]

**Review by axc450 — CHANGES_REQUESTED** [review]
I'm not sure about `cond` and `clamp`. Can we drop these for now and just implement the unary/binary operations? Maybe we should just start with the basic math operations.

**Comment by axc450** [comment]
A regression test script would also be nice for this

File: Core/Util.lua [hunk 1]
```diff
@@ -141,3 +141,42 @@ end
...
+    clamp = function(value, minClamp, maxClamp) return min(max(value, minClamp), maxClamp); end,
+}
+
+local mathParserSafeEnv = {}
+setmetatable(mathParserSafeEnv, { __index = mathParserEnv, __newindex = function() end, __metatable = false })
+
+local function mathParseExpression(expression)
+    forceinsecure()
+
+    local expr = loadstring("return ("..expression..")")
```
Comment (axc450) [inline Core/Util.lua:162]: I feel this is not the safest way to do this. I would prefer writing a manual parser. While it would take more time to do, we would have proper control over user input.

Comment (bloerwald) [reply Core/Util.lua:162]: Blizzard trusts it at least, then again, they have all input under control.

`loadstring()` doesn't execute anything yet. `setfenv()` *completely* replaces the environment. The only things available within that environment are things defined within that expression, and in the given table. So only `mathParserEnv` is available in there.

`__index` gives only those functions that are listed with no way to modify. `__newindex = function() end` prevents adding anything to that table from within. `__metatable` prevents changing the metatable using `setmetatable()`.

The internet suggests to not provide `getfenv()`, `getmetatable()` and `debug` inside the environment, which we don't. 

Tl;dr: This should be perfectly safe.

File: Core/Util.lua [hunk 2]
```diff
@@ -141,3 +141,42 @@ end
...
+
+-- See Interface/FrameXML/SharedPetBattleTemplates.lua's SharedPetAbilityTooltip_ParseExpressionInsecure for how this works.
+local mathParserEnv = {
+    ceil = math.ceil,
+    floor = math.floor,
+    abs = math.abs,
+    min = math.min,
+    max = math.max,
+    cond = function(conditional, onTrue, onFalse) if ( conditional ) then return onTrue; else return onFalse; end end,
+    clamp = function(value, minClamp, maxClamp) return min(max(value, minClamp), maxClamp); end,
```
Comment (bloerwald) [inline Core/Util.lua:153]:
```suggestion
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -3,6 +3,7 @@
 - Fixed issue with Test button not correctly disabling during pet battle animations
 - Fixed issue where the Test button crashed on if/endif blocks in the script.
 - Fixed condition (pet) `type` to correctly handle bad pets (e.g. `self(#4).type`) again.
+- Allow using math within "compare" conditions, like `hp` or `speed`. Basic math operations like `+-*/%()` are available. Additionally the functions `ceil()` and `floor()`, `abs()`, `min(a, b)` and `max(a, b)`, `cond(c, yes, no)` and `clamp(v, min, max)` can be used.
```
Comment (bloerwald) [inline CHANGELOG.md:6]:
```suggestion
- Allow using math within "compare" conditions, like `hp` or `speed`. Basic math operations like `+-*/%()` are available. Additionally the functions `ceil()` and `floor()`, `abs()`, `min(a, b)` and `max(a, b)` can be used.
```

**Comment by bloerwald** [comment]
> Would we want to return the floored (or rounded) value of the expression instead of the actual value? `(732 * 1.25) / 2 = 457.5` but I think in this case pets actually hit for `457`.
> 
> I guess this only matters in rare cases where the enemy pet has exactly 457 health, and I'm not sure if the actual attack value is floored or rounded to closest int. I think at least the 'hp' condition should consider this and deal with any flooring/rounding of the final value to match how blizzard calculates pet values?

Hm. Is floor always correct? It seems correct in your example, what about a condition `hp ≥ x`? There is an explicit `floor()` available, so (the probably rare) users can explicitly do it for their conditions? 

The Blizzard equivalent code does `floor()` always, but they only care about tooltip numbers not having too much detail..

**Comment by erigul** [comment]
> > Would we want to return the floored (or rounded) value of the expression instead of the actual value? `(732 * 1.25) / 2 = 457.5` but I think in this case pets actually hit for `457`.
> > I guess this only matters in rare cases where the enemy pet has exactly 457 health, and I'm not sure if the actual attack value is floored or rounded to closest int. I think at least the 'hp' condition should consider this and deal with any flooring/rounding of the final value to match how blizzard calculates pet values?
> 
> Hm. Is floor always correct? It seems correct in your example, what about a condition `hp ≥ x`? There is an explicit `floor()` available, so (the probably rare) users can explicitly do it for their conditions?
> 
> The Blizzard equivalent code does `floor()` always, but they only care about tooltip numbers not having too much detail..

I was thinking that for `hp` condition it would be nice if it was done by the addon. The value we are parsing is most likely an attack value, and if we do the correct handling of non-integer values script writes won't have to think about the correct way of doing it. Also if what blizzard actually does internally is rounding (not flooring) doing automatically might be the only way. 

But it's probably such a narrow edge case that it doesn't matter.

File: Core/Util.lua (hunk 1 above)
Comment (axc450) [reply Core/Util.lua:160]: Does it still allow arbitrary execution of LUA code that is placed in a pet battle script? I am convinced it is safe, so thanks for the explanation! 😄But it would be quite odd if we start seeing scripts that contain LUA which this line executes, even if safe.

Comment (erigul) [reply Core/Util.lua:160]: Should mathParseExpression check that what's returned is actually a number? That way even if people try to get up to funny business with LUA they shouldn't be able to return anything we don't expect.

**Comment by bloerwald** [comment]
Re floor/round: I was referring to [the interface code for toolitips](https://github.com/tomrus88/BlizzardInterfaceCode/blob/e4385aa29a69121b3a53850a8b2fcece9553892e/Interface/FrameXML/SharedPetBattleTemplates.lua#L426) where they floor numbers often, not round. That’s only Interfaxe though. 

I don’t think I’d add rounding automatically anywhere. You want to use math, make sure it makes sense!

File: Core/Util.lua (hunk 1 above)
Comment (bloerwald) [reply Core/Util.lua:160]: We can add that, I don’t have hard feelings. We currently only have this for mathy stuff but technically we could use it for other conditions as well where non-number return values make sense.

Just slap `tonumber()` around the last return?

**Comment by cosinussinus** [comment]
I have done a lot of pet battle math. Blizz does both: rounding and truncing. And on top of that Blizz doesn't care about floating point errors. In my experience we don't need math in tdscript.

**Comment by axc450** [comment]
Closing as discussed. I would probably only consider the basic arithmetic operations for now.
